Exceptions.md

```diff
@@ -0,0 +1,199 @@
+```
+
+which would enable us to call exception-throwing C# methods from Dafny without an extra unverified layer turning exceptions into `Result<int32>`.
+
+In the same way, methods implemented in Dafny which return `Result<T>` would be compiled to exception-throwing methods and thus be directly usable by C# client programs.
+
+Moreover,
+
+```
+return Failure(someErrorMsg);
+```
+
+is compiled into a `throw` statement:
+
+```
+throw new Exception(dafnyStringToNativeString(someErrorMsg));
+```
+
+Compilation of
+
+```
+return Success(someValue);
+```
+
+simply drops the wrapping:
+
+```
+return someValue;
+```
+
+And the new Dafny construct
+
+```
+var value1: T1 :- operation1(args1);
+```
+
+is simply compiled into
+
+```
+var value1 = operation1(args1);
+```
+
+and the handling of the failure case is done by C#'s exception mechanism.
+
+
+## FAQ
+
+*Q:* Are we also adding special syntax/support for this inside expressions?
+*A:* Not at the moment
+
+*Q:* What about target languages which do not have exceptions, such as Go?
+*A:* Go uses a pattern of returning an additional return value to report success/failure, and the Dafny-to-Go compiler would use this pattern too. In general, the idea is to compile `Result<T>` to the idiomatic way of dealing with failure in the target language, be it exceptions, extra return values, `Option`, `Maybe`, ...
+
+*Q:* Can this work for methods which have multiple return values?
+*A:* It seems so.
+
+
+## Discussion points, open questions
+
+### Should  `Result<T>` be first-class?
+
+Should it be possible to pass around values of type `Result<T>`? That is, should it be allowed to assign a `Result<T>` to a variable, or to pass a `Result<T>` to a method? Say the user writes
+```
+var res1: Result<T1> := operation1(args1);
+anotherMethodCall(res1);
+```
+If this happens, and the compiler compiled `operation1` into a `public T1 operation1(args) throws Exception`, then the code generated for the above snippet would have to catch the exception and materialize it into a `Result<T1>`.
+In most cases, this is probably not the code which the user meant to write, so the questions are: Should this still be supported? Should it emit a warning? Should it emit an error?
+
+
+### How to support returning through assignement instead of `return`?
+
+Should the following be supported?
+
+```
+method Test() returns (res: Result<int32>) {
+	res := someMethodCall();
+	if changedMyMind() {
+		res := someOtherMethodCall();
+	}
+}
+```
+
+The most straightforward way of supporting this would be to surround `someMethodCall` and `someOtherMethodCall` with a `try/catch` block, materialize their result into a `Result<int32>`, and at the end of the method body, to return `res.get` or throw `res.error`.
+This seems inefficient, can we do better?
+Should it be supported at all, or would we only allow `return` in methods returning `Result`, but no assignment to the result variable?
+
```

Comment:
I think it would be nice to support.

Would it be possible to never involve any exception throwing/handling (or the Go equivalent pattern) except for methods marked with `{:extern}` in some way? If so, then maybe it could be clear what to do in these cases.